backend/app/services/action_engine.py

```
@@ -0,0 +1,495 @@
+- No memoization — recomputed on every dashboard load. Cost is one pass
+  over already-loaded arrays.
+- Cap the output at 8 entries; the frontend hides past 3 behind
+  ``[Show N more]``.
+- Each action's ``id`` is ``"{action_id}.{subject_id}"`` so React keys are
+  stable.
+"""
+
+from __future__ import annotations
+
+from datetime import datetime, timezone
+from typing import Iterable
```

Ruff (0.15.12)

[warning] Import from `collections.abc` instead: `Iterable`

Import from `collections.abc`

(UP035)